### PR TITLE
Feature: Improved behavior when dragging items from an unfocused Files window

### LINKED ISSUE #13255: Feature: Don't focus Files window when dragging a file/folder
### What feature or improvement do you think would benefit Files?

When trying to drag a file (or folder) from a Files window to other window, the Files windows will be focused and automatically jump to the top, covering other window that was originally above. It will be better if the window could stay behind.

It is necessary for the app popping onto the top when clicking on it, or dragging on some empty spaces., title bar, etc.. However, it would be more convenient if it could stay behind when dragging on a single file or the files selected. (That's how windows files explorer behaves on this situation.)

### Requirements

- When dragging an item from an unfocused Files window, the window should not be focused, stay behind, until the item is finally drop to the original window.
- If some of the files have been selected, the window should stay behind when dragging on those files, but get to the top when dragging on other unselected files.

### Files Version

2.5.21.0

### Windows Version

10.0.22631.2262

### Comments

_No response_

### PR DESCRIPTION
**Pull request by Jack251970**
**Resolved / Related Issues**

- Added support for not focusing Files window when dragging a file/folder
- Closes #13255

**Steps used to test these changes**

Stability is a top priority for Files and all changes are required to go through testing before being merged into the repo. Please include a list of steps that you used to test this PR.

1. Opened Files and File Explorer.
2. Select files in Files.
3. Move File Explorer on the top, which is topper than Files. Make sure it doesn't cover Files so that you can drag files from Files.
4. Drag the selected files to the File Explorer. And you can find the Files won't move to the top when dragging files.

### CODE REVIEW

File: src/Files.App/Views/Layouts/BaseLayoutPage.cs
Comment (yaira2) [inline src/Files.App/Views/Layouts/BaseLayoutPage.cs:1342]: Can these statements be combined?

File: src/Files.App/Helpers/Win32/Win32Helper.WindowManagement.cs
Comment (yaira2) [inline src/Files.App/Helpers/Win32/Win32Helper.WindowManagement.cs:10]: We're trying to avoid new references to Vanara. While it's not absolutely required, it would be preferable if this can be achieved differently. Refer to issue #15000 for more details.

Comment (Jack251970) [reply src/Files.App/Helpers/Win32/Win32Helper.WindowManagement.cs]: Get it. Finished in 1aa2cd6.

**Review by yaira2 — DISMISSED** [review]

**Comment by yaira2** [comment]
@Jack251970 the behavior is working nicely when Files is maximized. Is there a way to also do this when Files isn't maximized?

**Comment by Jack251970** [comment]
@yaira2  Sorry, I cannot get your meaning. What is the difference between the behaviour when File is maximized and when it is not?

**Comment by Jack251970** [comment]
https://github.com/user-attachments/assets/91b17452-b706-47fe-be16-89460ec12508
I think the behaviour like this is enough compared to system File Explorer?

**Comment by yaira2** [comment]
> @yaira2 Sorry, I cannot get your meaning. What is the difference between the behaviour when File is maximized and when it is not?

I meant that Files is brought to the foreground if I drag a file when the Files window isn't maximized.

**Comment by Jack251970** [comment]
> > @yaira2 Sorry, I cannot get your meaning. What is the difference between the behaviour when File is maximized and when it is not?
> 
> I meant that Files is brought to the foreground if the window isn't maximized.

Sorry, still no idea about why Files needs to be brought to the foreground if the window isn't maximized. (Under what situation?)

As you can see in the video above, my codes implement a function that the Files will stay in the current z index if dragging selected files so that it will have the same experience as system File Explorer.

**Comment by 0x5bfa** [comment]
Putting aside the behavior, you can use DidPositionChanged from AppWindow.Changed event instead of utilizing win32 apis.

**Comment by Jack251970** [comment]
> Putting aside the behavior, you can use DidPositionChanged from AppWindow.Changed event instead of utilizing win32 apis.

I am not quite familiar with this api. Will it make some differences?

**Comment by yaira2** [comment]
@Jack251970 I'm referring to the behavior (around two seconds into your recording) where the Files window is brought to the foreground.
1. Open Files regularly (not maximized)
2. Select file
3. Switch focus to File Explorer
4. Start dragging selected file from Files
5. Files is brought to the foreground

Isn't the expected behavior for File Explorer to remain the topmost window?

**Comment by Jack251970** [comment]
> @Jack251970 I'm referring to the behavior (around two seconds into your recording) where the Files window is brought to the foreground.
> 
> 1. Open Files regularly (not maximized)
> 2. Select a file
> 3. Switch the focus to File Explorer
> 4. Start dragging the selected file from Files
> 5. Files is brought to the foreground
> 
> Isn't the expected behavior for File Explorer to remain the topmost window?

Get it. I will try to make this work.

**Comment by yaira2** [comment]
Thanks! I should note, your current changes already helped to resolve this issue when the window is maximized.

**Comment by 0x5bfa** [comment]
> Will it make some differences?

Yeah you don't have to hook/unhook win32 window message loop, there's high-level API wrapper called WinRT API as opposed to low-level Win32 API.

This API let you just subscribe an event. Go to MainWindow and subscribe this.Changed event and get e.DidPositionChanged property value.

**Comment by Jack251970** [comment]
> > Will it make some differences?
> 
> Yeah you don't have to hook/unhook win32 window message loop, there's high-level API wrapper called WinRT API as opposed to low-level Win32 API.
> 
> This API let you just subscribe an event. Go to MainWindow and subscribe this.Changed event and get e.DidPositionChanged property value.

Sure, I will have a try.

**Comment by Jack251970** [comment]
@yaira2 Sorry about the misunderstanding description (I have removed the useless step that asks you to maximize the Files window). This feature can work while Files isn't a maximized window.
In my recording around 2 seconds, the reason why Files is brought to the front is that I have selected a new item (09-20-37.mp4) instead of the original one (09-28-05.mp4). This behaviour is just like the system File Explorer.

**Comment by Jack251970** [comment]
@0x5bfa Yeah I have found that AppWindow.Changed event contains information about DidPositionChange & ZOrderChange, but  I cannot handle these changes. (AppWindowChangedEventArgs does not have Handle property, and I cannot stop those changing event)

So I think WindowMessageReceived event in WinUIEx.WindowManager is the only way to implement this and WinUIEx package has provided a good solution for this.

**Comment by yaira2** [comment]
@Jack251970 thank you for clarifying. I'll test again to confirm the behavior works on my side.

**Comment by 0x5bfa** [comment]
> Yeah I have found that AppWindow.Changed event contains information about DidPositionChange & ZOrderChange, but I cannot handle these changes. (AppWindowChangedEventArgs does not have Handle property, and I cannot stop those changing event)

Thanks, I missed that. I'll file a feature request on WinAppSdk for this later.

**Comment by yaira2** [comment]
@Jack251970 I can confirm the behavior is working as expected. I'm going to wait for a review from @hishitetsu, but as far as I'm concerned I have no requested changes.

**Comment by hishitetsu** [comment]
This improvement does not work once the app goes to the background and then reopens. 
I also found a few behaviors that differ from Explorer.
* In Files, if you try to drag a file that is not selected, the Files window comes into focus; this is not the case in Explorer.
* Once dragged to another window, returning the cursor and releasing the mouse button on the original window will focus the original window in Explorer, but not in Files.

**Review by hishitetsu — COMMENTED** [review]
This improvement does not work once the app goes to the background and then reopens. 
I also found a few behaviors that differ from Explorer.
* In Files, if you try to drag a file that is not selected, the Files window comes into focus; this is not the case in Explorer.
* Once dragged to another window, returning the cursor and releasing the mouse button on the original window will focus the original window in Explorer, but not in Files.

**Comment by Jack251970** [comment]
@hishitetsu Thank you for testing this function so carefully!

As for few behaviors that differ from Explorer above, I don not think it is necessary to ensure Files has the same behavior as the Explorer and I wonder what is the difference in user experience. (Sorry, I am not familiar with focus behaviors.)

For `This improvement does not work once the app goes to the background and then reopens`, do you mean close the Files window into tray icon? (I have test it when minimize File window into task bar, and I found it can work.)

**Comment by hishitetsu** [comment]
> can you provide the reproduction steps?

1. Set `Leave app running in the background when the window is closed` option to on.
2. Close (not minimize) the Files window.
3. Click the task tray icon to reopen Files.
4. Test the behavior.

**Comment by Jack251970** [comment]
@hishitetsu Now fixed, I think it can work when you close window and keep it in the background.

**Comment by yaira2** [comment]
> As for few behaviors that differ from Explorer above, I don not think it is necessary to ensure Files has the same behavior as the Explorer and I wonder what is the difference in user experience. (Sorry, I am not familiar with focus behaviors.)

For what it's worth, it's probably better if we only implement the behaviors as they are requested. Thanks for all your work!

**Review by yaira2 — APPROVED** [review]

File: src/Files.App/MainWindow.xaml.cs
Comment (hishitetsu) [inline src/Files.App/MainWindow.xaml.cs:353]: This does not guarantee atomic operation when multiple threads handle it at the same time. It is better to use `Interlocked` class or `lock` statement.

File: src/Files.App/MainWindow.xaml.cs
Comment (hishitetsu) [inline src/Files.App/MainWindow.xaml.cs:366]:
```suggestion
			if (!CanWindowToFront && e.Message.MessageId == WM_WINDOWPOSCHANGING)
			{
				Win32Helper.ForceWindowPosition(e.Message.LParam);
				e.Handled = true;
			}
```

**Review by hishitetsu — COMMENTED** [review]
It seems to work fine, but there are a few comments about the code quality.

**Comment by Jack251970** [comment]
@hishitetsu Resolved.

File: src/Files.App/MainWindow.xaml.cs
Comment (hishitetsu) [inline src/Files.App/MainWindow.xaml.cs:358]: Use tabs for indentation.
```suggestion
		public bool SetCanWindowToFront(bool canWindowToFront)
		{
			lock (_canWindowToFrontLock)
			{
				if (CanWindowToFront != canWindowToFront)
				{
					CanWindowToFront = canWindowToFront;
					return true;
				}
				return false;
			}
		}
```

**Review by hishitetsu — APPROVED** [review]
LGTM

File: src/Files.App.CsWin32/Windows.Win32.Extras.cs
Comment (0x5bfa) [inline src/Files.App.CsWin32/Windows.Win32.Extras.cs:24]: Don't add any to this file. Use CsWin32's one.

File: src/Files.App/MainWindow.xaml.cs
Comment (0x5bfa) [inline src/Files.App/MainWindow.xaml.cs:360]: Add this to CsWin32 and use it thru PInvoke.WM_WINDOWPOSCHANGING

File: src/Files.App/Views/Layouts/BaseLayoutPage.cs
Comment (0x5bfa) [inline src/Files.App/Views/Layouts/BaseLayoutPage.cs:1219]: Please add issue number like this.
There's also 5 places to add this below this line.
```suggestion
			// Set can window to front and bring the window to the front if necessary (#13255)
```

**Pull request merged**
